test/aspect/aspect.bzl

```diff
@@ -1,49 +1,56 @@
-attr_aspects = [
-    "deps",
-    "runtime_deps",
-    "_scalalib",
-    "_scalareflect",
-    "_scalatest_reporter",
-]
+"""
+This test makes sure that the implicit rule dependencies are discoverable by
+an IDE. We stuff all dependencies into _scala_toolchain so we just need to make
+sure the targets we expect are there.
+"""
+scala_toolchain = "_scala_toolchain"
 
 def _aspect_impl(target, ctx):
-    visited = [target.label.name]
-    for name in attr_aspects:
-        if hasattr(ctx.rule.attr, name):
-            attr = getattr(ctx.rule.attr, name)
-            # Need to handle whether attribute is label or label list
-            children = attr if type(attr) == "list" else [attr]
-            for child in children:
-                if hasattr(child, "visited"):
-                    visited += child.visited
+    visited = [str(target.label)]
+    if hasattr(ctx.rule.attr, scala_toolchain):
+        for dep in getattr(ctx.rule.attr, scala_toolchain):
+            if hasattr(dep, "visited"):
+                visited += dep.visited
     return struct(visited = visited)
 
 test_aspect = aspect(
-    attr_aspects = attr_aspects,
+    attr_aspects = [scala_toolchain],
     implementation = _aspect_impl,
 )
 
 def _rule_impl(ctx):
-    expected = [
-        "dummy",
-        "jar", # This is scalatest since @scalatest/jar
-        "scala-library",
-        "scala-reflect",
-        "scala-xml", # dependency of test_reporter
-        "test_reporter",
-    ]
-    # Remove duplicates and sort so can do simple comparison
-    visited = sorted(depset(ctx.attr.scala_rule.visited).to_list())
-    if visited == expected:
-        content = "true"
-    else:
-        content = """
-        echo Expected these rules to be visited by the aspect: 1>&2
-        echo %s, 1>&2
-        echo but got these instead: 1>&2
-        echo %s 1>&2
-        false
-        """ % (', '.join(expected), ', '.join(visited))
+    deps = {
+        "scala_library" : [
+            "//test/aspect:scala_library",
+            "@scala//:scala-library",
+        ],
+        "scala_test" : [
+            "//test/aspect:scala_test",
+            "@scala//:scala-library",
+            "@scalatest//jar:jar",
+        ],
+        "scala_junit_test" : [
+            "//test/aspect:scala_junit_test",
+            "@scala//:scala-library",
+            "@io_bazel_rules_scala_junit_junit//jar:jar",
+            "@io_bazel_rules_scala_org_hamcrest_hamcrest_core//jar:jar",
+        ],
+    }
+    content = ""
+    for rule in [ctx.attr.scala_library,
+                 ctx.attr.scala_test,
+                 ctx.attr.scala_junit_test]:
+        expected = deps[rule.label.name]
+        if rule.visited != expected:
+            content += """
+            echo Expected these deps from {name}: 1>&2
+            echo {expected}, 1>&2
+            echo but got these instead: 1>&2
+            echo {visited} 1>&2
+            false
```

Comment:
Does "false" make the test fail? if so I'd really like to learn why, if not what makes the test fail?